### ISSUE TITLE
rename / copy command

### ISSUE THREAD

**Issue by KutnerUri** [teambit/bit#2688]
Sometimes I am forced to rename a component, but Bit does not provide an easy to way to it.(unless I'm missing something?)

## Existing solutions
Currently, to rename I have to trick Bit into thinking the component is new, so I could tag it under the desired name.

I lose the component tag history and examples, and I have to manually go and update all consumers.
The closest thing available is `reset-to-add`, and `export` to a new scope (fork), both of which don't quite rename.

This is my process:

1. `bit remove <component-id>` (without `--remote` or `--delete-files`)
1. Rename folder / files.
1. `bit add <new-path> -n <namespace>` (or specifically with `--id`)
1. `bit tag <component-id>` (usually with `--skip-auto-tag`, or automatically with `--all`)
1. `bit export` (to default scope)
1. `bit deprecate --remote <old-component-id>`

It's not very pleasant, but it works, and it can be done in bulk.

## Goals
It seems there are 3 goals for rename:

* Preserving component history
* Delivering smooth updates
* Automating name change

## Suggested commands
```
bit copy (prevName) (nextName)
```

* clone an existing components to a new folder, with all its tags and data.

```
bit succeed (prevComponent) (nextComponent)
```

(I know, not the best name, but it describes it most accurately)

* Mark _prevComponent_ as deprecated
* Replace the content with a link to _nextComponent_, i.e. `export * from 'nextComponent';`. This should cover the basics, but the user should have an option to override this default (e.g. in case of scss files)
* Add a warning (during post-install?) to replace it with _nextComponent_.
* An option to update consumers would also be nice. (similar to `export --rewire`)

```
bit rename (prevName) (nextName)
```

* "copy" component (without changing files).
* mark as successor
* update bitmap
* (not a must) `rewire` consumers retroactively would be nice. → Could be done by `ripple.`

Ideally, the commands should not tag the existing component.

## Bulk
Of course, we will also need this as a bulk operation. (For example, for namespaces). This has its own caveats:

* We should show a preview to be approved by the user. (either with a opt-in `--dry-run` flag or an opt-out `-y` flag)
* We need a way to describe the change (perhaps with regex? or maybe stdin, like `bit ls | grep /sections/ | sed s#^/sections/#/sections/enterprise/#`)
* Some changes may only be possible in bulk, like how tag runs in toposortally.

**Comment by GiladShoham** [comment]
@KutnerUri Thanks for reporting this.
We should define a product for this.
Once we do, I'll update it here.

**Comment by KutnerUri** [comment]
open questions:
- should we keep the old history (versions, etc) when cloning?
  --> yes! exactly as is. You should emit a "clone" event for subscribers to run side effects (like copy screenshots, etc)
- what happens if variance change (workspace configurations) 
  --> show [y/N] warning, and specify what will change
- should this happen in a workspace, or can it happen also in a remote scope?
  --> make it work in local workspace first.
  making it work in remote scope should be easier, since there is no variances to consider, etc
- what do we do with `aspect` components?
  --> not critical, can block them
- what do we do with the old component
  --> deprecate it, and do not delete. Other components might still rely on it.
- what happens if the component is used in a template?
  --> it would still work. users will have to update manually.

**Comment by davidfirst** [comment]
Discussed this with @GiladShoham and @ranm8 , I'll summarize it here.

There are two, very similar scenarios. One is "rename". The other is "fork". Here are the steps to enable these features:

1. If a component is new, it's easy. Just rename it in the .bitmap file and that's it.
2. add a flag `--refactor-dependents-import`, to search & replace all import statements of the dependents in the user source code. (codemod)
3. If the component is exported and has dependents. a new component will be created and once tagged, it'll save a reference to the original component. In case of rename, the original component will be marked as deprecated and it'll save a reference as well to the new component, so then on import we can suggest users to use the new component. In case of fork, it won't touch the original component (the user might not have write permissions to the original component).
4. If the component is exported and doesn't have any dependents (we should verify it in bit-dev against other scopes as well), the old component can be removed in case of "rename", and a new component will be created. (open questions about this scenario: do we want to deprecate the old one to keep history? do we want to keep references to the old one?)

In regard to the deprecation. 
currently, to deprecate a remote component, you must use `bit deprecate --remote` . If you deprecate locally and export, it doesn't change anything on the remote. 
When renaming a component, we want behind the scenes to deprecate the original component locally, and shows it as staged/export-pending on `bit status`, so then `bit export` will push the changes to the remote. 
For this to work, we should first change the mechanism of `bit deprecate` to work in the same way. In terms of implementation, it can work similar to adding versions. We can add a new prop to `ComponentModel.state`, to indicate whether a component was deprecated locally. Once it's there, the component is "staged" and it'll be export pending.
Another option is to create a new "model" which saves this data. This way, it'll help enabling rename when on a lane. This is not finalized yet and will need to be determined.

**Comment by KutnerUri** [comment]
sounds good. We can show a deprecation warning as a post install script (or something similar), or maybe deprecate the package in `npm` (or any registry the package is configured to)

Ideally, the rename should be seamless, like
```
old/component
|
1.3.5
|
1.3.6
| 
1.3.7 
RENAME
  \
   \
    \
     \
      new component
       1.3.7
       |
       1.3.8
```

so, these things will be nice to have:
- ability to replace the old component with the new one with 0 risk (ie. new component should include the last tag of the previous one)
- fork the entire history (either as actual snaps, or as a readonly clone of the original component), so users will have a seamless experience in `bit log`, etc.
- in relevant commands, suggest replacing the old component (bit status, install, and others)
- provide an automatic script (either built in or 3rd party) that will find and update renamed components (with a readonly mode)

**Comment by itaymendel** [comment]
resolved.

```
bit rename
```

also,

```
bit fork
```
